### PR TITLE
[REVIEW] HighLevelGraph validation and dependencies fixes

### PR DESCRIPTION
**Pull request by madsbk**
Expanding on the work of #6509, this PR implement a complete HighLevelGraph validation by[ re-calculating the layer dependencies](https://github.com/madsbk/dask/blob/01942e16f2a0eb67c8e1ec08221eacdc117c8d72/dask/highlevelgraph.py#L11) and fixes all incorrect dependencies.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/delayed.py [hunk 1]
```diff
@@ -470,21 +470,31 @@ class Delayed(DaskMethodsMixin, OperatorMethodMixin):
...
 
     def __dask_keys__(self):
         return [self.key]
 
     def __dask_layers__(self):
-        return (self.key,)
+        if hasattr(self, "_dask_layers"):
+            return self._dask_layers
+        else:
+            return (self.key,)
```
Comment (mrocklin) [inline dask/delayed.py:497]: I would expect this to always be `self.key`.  I'm curious in which cases this assumption is incorrect.

**Comment by mrocklin** [comment]
In general this seems like a very useful cleanup.  Thank you @madsbk for doing this.  I am curious about `Delayed.__dask_layers__` but otherwise everything here seems straightforward.

File: dask/delayed.py (hunk 1 above)
Comment (madsbk) [reply dask/delayed.py:497]: This boils down to the discrepancy between the definition of `__dask_layers__` that should return _output layers_ and `Delayed` that might not have any layers. 
`Delayed.key` is always a _low level_ graph key and so this is kind of a hack that make `Delayed` return a layer if possible.

**Comment by mrocklin** [comment]
I apologize for the delay in reviewing this.  This looks good to me.  Thank you for this cleanup @madsbk .  Merging.

**Pull request merged**